### PR TITLE
Save proofs

### PR DESCRIPTION
**Pull request by SimonGuilloud**
Tool to export proofs of theorems to binary files and back.

### CODE REVIEW

**Comment by SimonGuilloud** [comment]
- Add ids to terms like for formulas
- Add serialization of proofs. Theorems can be serialized to binary files and loaded back on later run. 
- On my machine, running everything up to Recursion.scala takes 19s instead of 24s.
- Does no hash consing, but simple optimizations to proof size (remove consecutive rewrites, flatten subproofs)
- Cleans up the distinction between fullName (with the whole path, should be unique) and name (just the last part, possibly duplicate across different files/domains).
- Good completion of documentation in WithTheorems
- checkProofs does not print proofs of more thant 100 steps now.
- fixed an indexing bug in ShrinkProof.flattenProof
- Suite of tests for serialization, export then load a collection of theorems.
- Push suite of tests for Tableaux tactic that were missing.

File: lisa-utils/src/main/scala/lisa/prooflib/Library.scala [hunk 1]
```diff
@@ -30,6 +30,8 @@ abstract class Library extends lisa.prooflib.WithTheorems with lisa.prooflib.Pro
 
   var last: Option[JUSTIFICATION] = None
 
+  var _withCache: Boolean = false
```
[the remark this answers is not in the document]
Comment (sankalpgambhir) [reply lisa-utils/src/main/scala/lisa/prooflib/Library.scala:33]: Not in this PR necessarily, but going forward.

[the remark this answers is not in the document]
Comment (SimonGuilloud) [reply lisa-utils/src/main/scala/lisa/prooflib/Library.scala:33]: Good idea, I'll do that somewhat soon, probably in the next PR

File: src/main/scala/lisa/settheory/SetTheoryZAxioms.scala [hunk 2]
```diff
@@ -19,7 +19,7 @@ private[settheory] trait SetTheoryZAxioms extends SetTheoryDefinitions {
    *
    * `() |- (x = y) ⇔ ∀ z. z ∈ x ⇔ z ∈ y`
    */
-  final val extensionalityAxiom: this.AXIOM = Axiom("extensionalityAxiom", forall(z, in(z, x) <=> in(z, y)) <=> (x === y))
+  final val extensionalityAxiom: this.AXIOM = Axiom(forall(z, in(z, x) <=> in(z, y)) <=> (x === y))
```
Comment (sankalpgambhir) [inline src/main/scala/lisa/settheory/SetTheoryZAxioms.scala:22]:
```suggestion
  final val extensionalityAxiom: AXIOM = Axiom(forall(z, in(z, x) <=> in(z, y)) <=> (x === y))
```

The type annotations in this file are slightly inconsistent, with just this one being different afaict

File: lisa-utils/src/main/scala/lisa/utils/Serialization.scala [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+  inline def weakening: Byte = 20
+  inline def leftRefl: Byte = 21
+  inline def rightRefl: Byte = 22
+  inline def leftSubstEq: Byte = 23
+  inline def rightSubstEq: Byte = 24
+  inline def leftSubstIff: Byte = 25
+  inline def rightSubstIff: Byte = 26
+  inline def instSchema: Byte = 27
+  inline def scSubproof: Byte = 28
+  inline def sorry: Byte = 29
```
Comment (sankalpgambhir) [inline lisa-utils/src/main/scala/lisa/utils/Serialization.scala:41]: Wouldn't it be better to have a `toByte` with pattern matching? That way we have pattern match exhaustivity checking?

**Review by sankalpgambhir — APPROVED** [review]
Looks great overall. Minor optional changes as above ^^ 

Thanks for adding tests and documentation for adjacent files as well.

**Pull request merged**